src/plan/immix/global.rs

```diff
@@ -160,4 +162,37 @@ impl<VM: VMBinding> Immix<VM> {
 
         immix
     }
+
+    /// Schedule a full heap immix collection. This method is used by immix/genimmix/stickyimmix
+    /// to schedule a full heap collection. A plan must call set_collection_kind and set_gc_status before this method.
+    pub(crate) fn schedule_immix_full_heap_collection<
+        FastContext: 'static + GCWorkContext<VM = VM>,
+        DefragContext: 'static + GCWorkContext<VM = VM>,
+    >(
+        plan: &'static DefragContext::PlanType,
+        immix_space: &ImmixSpace<VM>,
+        scheduler: &GCWorkScheduler<VM>,
+    ) {
+        let in_defrag = immix_space.decide_whether_to_defrag(
+            plan.is_emergency_collection(),
+            true,
+            plan.base().cur_collection_attempts.load(Ordering::SeqCst),
+            plan.base().is_user_triggered_collection(),
+            *plan.base().options.full_heap_system_gc,
+        );
+
+        if in_defrag {
+            scheduler.schedule_common_work::<DefragContext>(plan);
+        } else {
+            // The type of plan is `DefragContext::PlanType`, and we need it as `FastContext::PlanType`.
+            // They should be the same plan. But I don't find a way to tell Rust compiler that
+            // those PlanTypes are the same. So just do a unsafe transmute here.
+            let fast_plan = unsafe { std::mem::transmute(plan) };
+            scheduler.schedule_common_work::<FastContext>(fast_plan);
```

Comment:
We can tell Rust that both `FastContext` and `DefragContext` has the same `PlanType` by adding an additional type parameter.  Here is a patch:

[same-type.patch.gz](https://github.com/mmtk/mmtk-core/files/10729156/same-type.patch.gz)

Reply:
Thanks. This is very helpful. I will make this change.